File: backend/routes/entries.py
"""
Journal entry endpoints -- CRUD, search, readings, follow-ups, tags, import/export.
Also serves profiles and spreads needed by the journal editor.
"""

import json
from flask import Blueprint, jsonify, request, current_app
from backend.services.richtext import convert_content_to_html
from backend.utils import row_to_dict

entries_bp = Blueprint('entries', __name__)

# Pagination limits to prevent memory exhaustion
MAX_LIMIT = 500
DEFAULT_LIMIT = 50


def _parse_cards_used(cards_json):
    """Parse cards_used JSON, handling both string-array and object-array formats."""
    if not cards_json:
        return []
    try:
        cards = json.loads(cards_json) if isinstance(cards_json, str) else cards_json
    except (json.JSONDecodeError, TypeError):
        return []

    result = []
    for card in cards:
        if isinstance(card, str):
            result.append({'name': card, 'reversed': False})
        elif isinstance(card, dict):
            result.append(card)
    return result


def _enrich_cards_with_ids(db, cards):
    """
    Enrich card data for display. Uses card_id as primary lookup (survives renames),
    falls back to deck_id + name for legacy entries without card_id.
    """
    # Separate cards by lookup strategy
    cards_with_id = []
    cards_needing_name_lookup = []

    for card in cards:
        if card.get('card_id'):
            cards_with_id.append(card['card_id'])
        else:
            deck_id = card.get('deck_id')
            name = card.get('name')
            if deck_id and name:
                cards_needing_name_lookup.append((deck_id, name))

    # Batch lookup cards by ID (to verify they exist and get current name)
    id_to_card = {}
    if cards_with_id:
        placeholders = ','.join('?' * len(cards_with_id))
        rows = db.conn.execute(
            f'SELECT id, deck_id, name FROM cards WHERE id IN ({placeholders})',
            cards_with_id
        ).fetchall()
        id_to_card = {row['id']: row for row in rows}

    # Batch lookup cards by (deck_id, name) for legacy entries
    name_to_id = {}
    if cards_needing_name_lookup:
        conditions = ' OR '.join(['(deck_id = ? AND name = ?)'] * len(cards_needing_name_lookup))
        params = [val for pair in cards_needing_name_lookup for val in pair]
        rows = db.conn.execute(
            f'SELECT id, deck_id, name FROM cards WHERE {conditions}',
            params
        ).fetchall()
        name_to_id = {(row['deck_id'], row['name']): row['id'] for row in rows}

    # Apply lookups to cards
    for card in cards:
        card_id = card.get('card_id')
        if card_id:
            # Card has ID - verify it exists and update name to current value
            db_card = id_to_card.get(card_id)
            if db_card:
                # Card still exists - use current name from database
                card['current_name'] = db_card['name']
            # If card was deleted, keep original data but card_id won't resolve to thumbnail
        else:
            # Legacy entry without card_id - try name lookup
            deck_id = card.get('deck_id')
            name = card.get('name')
            if deck_id and name:
                found_id = name_to_id.get((deck_id, name))
                if found_id:
                    card['card_id'] = found_id

    return cards


# ── Entries CRUD ──────────────────────────────────────────────

@entries_bp.route('/api/entries')
def list_entries():
    db = current_app.config['DB']
    limit = request.args.get('limit', DEFAULT_LIMIT, type=int)
    offset = request.args.get('offset', 0, type=int)
    # Clamp values to prevent memory exhaustion
    limit = max(1, min(limit, MAX_LIMIT))
    offset = max(0, offset)
    rows = db.get_entries(limit=limit, offset=offset)
    return jsonify([row_to_dict(r) for r in rows])


@entries_bp.route('/api/entries/search')
def search_entries():
    db = current_app.config['DB']
    tag_ids_raw = request.args.get('tag_ids', '')
    tag_ids = None
    if tag_ids_raw:
        try:
            tag_ids = [int(x) for x in tag_ids_raw.split(',') if x.strip()]
        except ValueError:
            return jsonify({'error': 'Invalid tag_ids format - must be comma-separated integers'}), 400

    rows = db.search_entries(
        query=request.args.get('query') or None,
        tag_ids=tag_ids,
        deck_id=request.args.get('deck_id', type=int),
        spread_id=request.args.get('spread_id', type=int),
        cartomancy_type=request.args.get('cartomancy_type') or None,
        card_name=request.args.get('card_name') or None,
        date_from=request.args.get('date_from') or None,
        date_to=request.args.get('date_to') or None,
    )
    return jsonify([row_to_dict(r) for r in rows])


@entries_bp.route('/api/entries/<int:entry_id>')
def get_entry(entry_id):
    """Return a fully hydrated entry with readings, tags, follow-ups, and profile names."""
    db = current_app.config['DB']
    row = db.get_entry(entry_id)
    if not row:
        return jsonify({'error': 'Entry not found'}), 404

    entry = row_to_dict(row)

    # Convert content to HTML
    entry['content'] = convert_content_to_html(entry.get('content'))

    # Readings with parsed + enriched cards
    readings = db.get_entry_readings(entry_id)
    entry['readings'] = []
    for r in readings:
        rd = row_to_dict(r)
        rd['cards_used'] = _enrich_cards_with_ids(db, _parse_cards_used(rd.get('cards_used')))
        entry['readings'].append(rd)

    # Tags
    tags = db.get_entry_tags(entry_id)
    entry['tags'] = [row_to_dict(t) for t in tags]

    # Follow-up notes (with HTML conversion)
    notes = db.get_follow_up_notes(entry_id)
    entry['follow_up_notes'] = []
    for n in notes:
        nd = row_to_dict(n)
        nd['content'] = convert_content_to_html(nd.get('content'))
        entry['follow_up_notes'].append(nd)

    # Querents (multiple) - from junction table
    querents = db.get_entry_querents(entry_id)
    entry['querents'] = [row_to_dict(q) for q in querents]

    # Legacy single querent name (for backwards compatibility)
    if entry.get('querent_id'):
        q = db.get_profile(entry['querent_id'])
        entry['querent_name'] = q['name'] if q else None
    else:
        entry['querent_name'] = None

    if entry.get('reader_id'):
        r = db.get_profile(entry['reader_id'])
        entry['reader_name'] = r['name'] if r else None
    else:
        entry['reader_name'] = None

    return jsonify(entry)


@entries_bp.route('/api/entries', methods=['POST'])
def create_entry():
    db = current_app.config['DB']
    data = request.get_json()
    if data is None:
        return jsonify({'error': 'Invalid or missing JSON body'}), 400
    entry_id = db.add_entry(
        title=data.get('title'),
        content=data.get('content'),
        reading_datetime=data.get('reading_datetime'),
        location_name=data.get('location_name'),
        location_lat=data.get('location_lat'),
        location_lon=data.get('location_lon'),
        querent_id=data.get('querent_id'),
        reader_id=data.get('reader_id'),
    )
    return jsonify({'id': entry_id}), 201


@entries_bp.route('/api/entries/<int:entry_id>', methods=['PUT'])
def update_entry(entry_id):
    db = current_app.config['DB']
    data = request.get_json()
    if data is None:
        return jsonify({'error': 'Invalid or missing JSON body'}), 400
    db.update_entry(
        entry_id,
        title=data.get('title'),
        content=data.get('content'),
        reading_datetime=data.get('reading_datetime'),
        location_name=data.get('location_name'),
        location_lat=data.get('location_lat'),
        location_lon=data.get('location_lon'),
        querent_id=data.get('querent_id'),
        reader_id=data.get('reader_id'),
    )
    return jsonify({'ok': True})


@entries_bp.route('/api/entries/<int:entry_id>', methods=['DELETE'])
def delete_entry(entry_id):
    db = current_app.config['DB']
    db.delete_entry(entry_id)
    return jsonify({'ok': True})


# ── Readings ──────────────────────────────────────────────────

@entries_bp.route('/api/entries/<int:entry_id>/readings')
def get_entry_readings(entry_id):
    db = current_app.config['DB']
    rows = db.get_entry_readings(entry_id)
    result = []
    for r in rows:
        rd = row_to_dict(r)
        rd['cards_used'] = _enrich_cards_with_ids(db, _parse_cards_used(rd.get('cards_used')))
        result.append(rd)
    return jsonify(result)


@entries_bp.route('/api/entries/<int:entry_id>/readings', methods=['POST'])
def add_entry_reading(entry_id):
    db = current_app.config['DB']
    data = request.get_json()
    if data is None:
        return jsonify({'error': 'Invalid or missing JSON body'}), 400
    reading_id = db.add_entry_reading(
        entry_id,
        spread_id=data.get('spread_id'),
        spread_name=data.get('spread_name'),
        deck_id=data.get('deck_id'),
        deck_name=data.get('deck_name'),
        cartomancy_type=data.get('cartomancy_type'),
        cards_used=data.get('cards_used'),
        position_order=data.get('position_order', 0),
    )
    return jsonify({'id': reading_id}), 201


@entries_bp.route('/api/entries/<int:entry_id>/readings', methods=['DELETE'])
def delete_entry_readings(entry_id):
    db = current_app.config['DB']
    db.delete_entry_readings(entry_id)
    return jsonify({'ok': True})


# ── Follow-up Notes ───────────────────────────────────────────

@entries_bp.route('/api/entries/<int:entry_id>/follow-up-notes')
def get_follow_up_notes(entry_id):
    db = current_app.config['DB']
    rows = db.get_follow_up_notes(entry_id)
    result = []
    for n in rows:
        nd = row_to_dict(n)
        nd['content'] = convert_content_to_html(nd.get('content'))
        result.append(nd)
    return jsonify(result)


@entries_bp.route('/api/entries/<int:entry_id>/follow-up-notes', methods=['POST'])
def add_follow_up_note(entry_id):
    db = current_app.config['DB']
    data = request.get_json()
    if data is None:
        return jsonify({'error': 'Invalid or missing JSON body'}), 400
    content = data.get('content', '')
    note_id = db.add_follow_up_note(entry_id, content)
    return jsonify({'id': note_id}), 201


@entries_bp.route('/api/follow-up-notes/<int:note_id>', methods=['PUT'])
def update_follow_up_note(note_id):
    db = current_app.config['DB']
    data = request.get_json()
    if data is None:
        return jsonify({'error': 'Invalid or missing JSON body'}), 400
    db.update_follow_up_note(note_id, data.get('content', ''))
    return jsonify({'ok': True})


@entries_bp.route('/api/follow-up-notes/<int:note_id>', methods=['DELETE'])
def delete_follow_up_note(note_id):
    db = current_app.config['DB']
    db.delete_follow_up_note(note_id)
    return jsonify({'ok': True})


# ── Entry Tags ────────────────────────────────────────────────

@entries_bp.route('/api/entries/<int:entry_id>/tags')
def get_entry_tags(entry_id):
    db = current_app.config['DB']
    rows = db.get_entry_tags(entry_id)
    return jsonify([row_to_dict(r) for r in rows])


@entries_bp.route('/api/entries/<int:entry_id>/tags', methods=['PUT'])
def set_entry_tags(entry_id):
    db = current_app.config['DB']
    data = request.get_json()
    if data is None:
        return jsonify({'error': 'Invalid or missing JSON body'}), 400
    tag_ids = data.get('tag_ids', [])
    db.set_entry_tags(entry_id, tag_ids)
    return jsonify({'ok': True})


# ── Entry Querents ─────────────────────────────────────────────

@entries_bp.route('/api/entries/<int:entry_id>/querents')
def get_entry_querents(entry_id):
    db = current_app.config['DB']
    rows = db.get_entry_querents(entry_id)
    return jsonify([row_to_dict(r) for r in rows])


@entries_bp.route('/api/entries/<int:entry_id>/querents', methods=['PUT'])
def set_entry_querents(entry_id):
    db = current_app.config['DB']
    data = request.get_json()
    if data is None:
        return jsonify({'error': 'Invalid or missing JSON body'}), 400
    profile_ids = data.get('profile_ids', [])
    db.set_entry_querents(entry_id, profile_ids)
    return jsonify({'ok': True})


# ── Profiles ──────────────────────────────────────────────────

@entries_bp.route('/api/profiles')
def get_profiles():
    db = current_app.config['DB']
    rows = db.get_profiles()
    return jsonify([row_to_dict(r) for r in rows])


@entries_bp.route('/api/profiles/<int:profile_id>')
def get_profile(profile_id):
    db = current_app.config['DB']
    row = db.get_profile(profile_id)
    if not row:
        return jsonify({'error': 'Profile not found'}), 404
    return jsonify(row_to_dict(row))


@entries_bp.route('/api/profiles', methods=['POST'])
def add_profile():
    db = current_app.config['DB']
    data = request.get_json()
    if data is None:
        return jsonify({'error': 'Invalid or missing JSON body'}), 400
    name = data.get('name', '').strip()
    if not name:
        return jsonify({'error': 'name is required'}), 400
    profile_id = db.add_profile(
        name=name,
        gender=data.get('gender'),
        birth_date=data.get('birth_date'),
        birth_time=data.get('birth_time'),
        birth_place_name=data.get('birth_place_name'),
        birth_place_lat=data.get('birth_place_lat'),
        birth_place_lon=data.get('birth_place_lon'),
        querent_only=data.get('querent_only', False),
        hidden=data.get('hidden', False),
    )
    return jsonify({'id': profile_id}), 201


@entries_bp.route('/api/profiles/<int:profile_id>', methods=['PUT'])
def update_profile(profile_id):
    db = current_app.config['DB']
    data = request.get_json()
    if data is None:
        return jsonify({'error': 'Invalid or missing JSON body'}), 400
    db.update_profile(
        profile_id,
        name=data.get('name'),
        gender=data.get('gender'),
        birth_date=data.get('birth_date'),
        birth_time=data.get('birth_time'),
        birth_place_name=data.get('birth_place_name'),
        birth_place_lat=data.get('birth_place_lat'),
        birth_place_lon=data.get('birth_place_lon'),
        querent_only=data.get('querent_only'),
        hidden=data.get('hidden'),
    )
    return jsonify({'ok': True})


@entries_bp.route('/api/profiles/<int:profile_id>', methods=['DELETE'])
def delete_profile(profile_id):
    db = current_app.config['DB']
    db.delete_profile(profile_id)
    return jsonify({'ok': True})


# ── Export / Import ───────────────────────────────────────────

@entries_bp.route('/api/entries/export')
def export_entries():
    """Export entries as JSON. Optional ?ids=1,2,3 to export specific entries."""
    db = current_app.config['DB']
    ids_raw = request.args.get('ids', '')
    entry_ids = None
    if ids_raw:
        try:
            entry_ids = [int(x) for x in ids_raw.split(',') if x.strip()]
        except ValueError:
            return jsonify({'error': 'Invalid ids format - must be comma-separated integers'}), 400
    data = db.export_entries_json(entry_ids)
    return jsonify(data)


@entries_bp.route('/api/entries/import', methods=['POST'])
def import_entries():
    """Import entries from JSON data."""
    db = current_app.config['DB']
    data = request.get_json()
    if data is None:
        return jsonify({'error': 'Invalid or missing JSON body'}), 400
    merge_tags = data.get('merge_tags', True)
    entries_data = data.get('data', data)
    result = db.import_entries_from_json(entries_data, merge_tags=merge_tags)
    return jsonify(result)

File: frontend/src/api/entries.ts
import api, { API_BASE } from './client';
import type {
  JournalEntry, JournalEntryFull, EntryReadingParsed,
  FollowUpNote, Tag, Profile,
} from '../types';

// ── Entries ──────────────────────────────────────────────────

export async function getEntries(limit = 50, offset = 0): Promise<JournalEntry[]> {
  const res = await api.get('/api/entries', { params: { limit, offset } });
  return res.data;
}

export async function getEntry(entryId: number): Promise<JournalEntryFull> {
  const res = await api.get(`/api/entries/${entryId}`);
  return res.data;
}

export async function searchEntries(params: {
  query?: string;
  tag_ids?: number[];
  deck_id?: number;
  spread_id?: number;
  cartomancy_type?: string;
  card_name?: string;
  date_from?: string;
  date_to?: string;
}): Promise<JournalEntry[]> {
  const p: Record<string, string> = {};
  if (params.query) p.query = params.query;
  if (params.tag_ids?.length) p.tag_ids = params.tag_ids.join(',');
  if (params.deck_id) p.deck_id = String(params.deck_id);
  if (params.spread_id) p.spread_id = String(params.spread_id);
  if (params.cartomancy_type) p.cartomancy_type = params.cartomancy_type;
  if (params.card_name) p.card_name = params.card_name;
  if (params.date_from) p.date_from = params.date_from;
  if (params.date_to) p.date_to = params.date_to;
  const res = await api.get('/api/entries/search', { params: p });
  return res.data;
}

export async function createEntry(data: {
  title?: string;
  content?: string;
  reading_datetime?: string;
  location_name?: string;
  location_lat?: number;
  location_lon?: number;
  querent_id?: number | null;
  reader_id?: number | null;
}): Promise<{ id: number }> {
  const res = await api.post('/api/entries', data);
  return res.data;
}

export async function updateEntry(entryId: number, data: {
  title?: string;
  content?: string;
  reading_datetime?: string;
  location_name?: string;
  location_lat?: number;
  location_lon?: number;
  querent_id?: number | null;
  reader_id?: number | null;
}): Promise<void> {
  await api.put(`/api/entries/${entryId}`, data);
}

export async function deleteEntry(entryId: number): Promise<void> {
  await api.delete(`/api/entries/${entryId}`);
}

// ── Readings ─────────────────────────────────────────────────

export async function getEntryReadings(entryId: number): Promise<EntryReadingParsed[]> {
  const res = await api.get(`/api/entries/${entryId}/readings`);
  return res.data;
}

export async function addEntryReading(entryId: number, data: {
  spread_id?: number | null;
  spread_name?: string;
  deck_id?: number | null;
  deck_name?: string;
  cartomancy_type?: string;
  cards_used?: Array<{ name: string; reversed?: boolean; deck_id?: number; deck_name?: string; position_index?: number }>;
  position_order?: number;
}): Promise<{ id: number }> {
  const res = await api.post(`/api/entries/${entryId}/readings`, data);
  return res.data;
}

export async function deleteEntryReadings(entryId: number): Promise<void> {
  await api.delete(`/api/entries/${entryId}/readings`);
}

// ── Follow-up Notes ──────────────────────────────────────────

export async function getFollowUpNotes(entryId: number): Promise<FollowUpNote[]> {
  const res = await api.get(`/api/entries/${entryId}/follow-up-notes`);
  return res.data;
}

export async function addFollowUpNote(entryId: number, content: string): Promise<{ id: number }> {
  const res = await api.post(`/api/entries/${entryId}/follow-up-notes`, { content });
  return res.data;
}

export async function updateFollowUpNote(noteId: number, content: string): Promise<void> {
  await api.put(`/api/follow-up-notes/${noteId}`, { content });
}

export async function deleteFollowUpNote(noteId: number): Promise<void> {
  await api.delete(`/api/follow-up-notes/${noteId}`);
}

// ── Entry Tags ───────────────────────────────────────────────

export async function getEntryTags(entryId: number): Promise<Tag[]> {
  const res = await api.get(`/api/entries/${entryId}/tags`);
  return res.data;
}

export async function setEntryTags(entryId: number, tagIds: number[]): Promise<void> {
  await api.put(`/api/entries/${entryId}/tags`, { tag_ids: tagIds });
}

// ── Entry Querents ──────────────────────────────────────────

export async function setEntryQuerents(entryId: number, profileIds: number[]): Promise<void> {
  await api.put(`/api/entries/${entryId}/querents`, { profile_ids: profileIds });
}

// ── Profiles ─────────────────────────────────────────────────

export async function getProfiles(): Promise<Profile[]> {
  const res = await api.get('/api/profiles');
  return res.data;
}

// ── Export / Import ──────────────────────────────────────────

export function exportEntriesUrl(entryIds?: number[]): string {
  const params = entryIds?.length ? `?ids=${entryIds.join(',')}` : '';
  return `${API_BASE}/api/entries/export${params}`;
}

export async function importEntries(data: unknown, mergeTags = true): Promise<{
  imported: number;
  skipped: number;
  tags_created: number;
}> {
  const res = await api.post('/api/entries/import', { data, merge_tags: mergeTags });
  return res.data;
}
